### ISSUE TITLE
Image with the same url loaded multiple times

### ISSUE THREAD

**Issue by vovkab** [bumptech/glide#359]
Hello

If you have url and try to load it multiple times with a different view size, image will not be loaded from cache, instead from network again and again.

My guess would be glide uses "size" as part of the key, which leads to a new key every time we have different view with the same url.

Here is example app:
https://github.com/vovkab/glide-example

**Comment by sjudd** [comment]
Ah yes your guess is exactly right. We're going to change this in Glide 4.0, but in Glide 3.0 we default to [``DiskCacheStrategy.RESULT``][1], which caches decoded, downsampled, and transformed data using the url, width, height, transformation id and various other ids in the cache key.

The alternative is to use [``DiskCacheStrategy.SOURCE``][2], which writes the retrieved data directly to disk without modification and uses only the original url as the cache key. You can set the disk cache strategy per request using [``diskCacheStrategy()``][3].

Since SOURCE is usually what you want for remote images, Glide 4.0 is going to change the defaults so that SOURCE is used for remote content and RESULT is used for local content. Unfortunately you will have to set the disk cache strategy manually in the meantime.

[1]: http://bumptech.github.io/glide/javadocs/350/com/bumptech/glide/load/engine/DiskCacheStrategy.html#RESULT
[2]: http://bumptech.github.io/glide/javadocs/350/com/bumptech/glide/load/engine/DiskCacheStrategy.html#SOURCE
[3]: http://bumptech.github.io/glide/javadocs/350/com/bumptech/glide/DrawableRequestBuilder.html#diskCacheStrategy(com.bumptech.glide.load.engine.DiskCacheStrategy)

**Comment by vovkab** [comment]
@sjudd thank you for a quick answer!

Do you know if there any way to set `diskCacheStrategy()` globally?

**Comment by vovkab** [comment]
Setting `.diskCacheStrategy(DiskCacheStrategy.SOURCE)` helped to avoid image reloads.

But I still see image start loading and get dropped pretty fast and start loading again, sometimes 3-4 tries. Sometimes glide start loading the same url 3 times at the same time, and then some where half the way drops others, sometimes it even loads 2-3 images for the same url.

But when it finish loading no more drops, looks like this is where `diskCacheStrategy()` kicks in.

p.s. Is it ok to keep it here or is it better to create a new ticket?

**Comment by sjudd** [comment]
Unfortunately there isn't any way to set disk cache strategy or other options globally, that's coming in Glide 4.

It's not totally surprising that a few requests might fail due to timeouts. You can see the code where most of the connection related settings are set here: https://github.com/bumptech/glide/blob/master/library/src/main/java/com/bumptech/glide/load/data/HttpUrlFetcher.java. Improvements are always welcome.

Similarly although Glide will batch identical requests within Glide, Glide makes no special effort to batch identical network requests. Two requests for a center cropped 100x100 pixel image at a given url will only cause the image to be decoded once, but a 100x100 request and a 200x00 request for a given url will cause the image to be decoded twice. 

I'd be willing to consider improving this for Glide's default network stack, though if you're looking for support now I'd suggest taking a look at a higher level library like Volley that will do this for you (see [the integration libraries wiki][1]).

[1]: https://github.com/bumptech/glide/wiki/Integration-Libraries

**Comment by vovkab** [comment]
I did more tests, and it looks like it is not completely a timeout issue.

Here is my theory, please correct me if I'm wrong :)
 If we have multiple views that need the same image, what will happens is:
1. we start first request
2. now new image view needs exactly the same image(url), so this new request will stop request1 and start a new one instead of waiting for request1
3. if we have 3-4 views like this or we set the same image multiple times, we would cancel and start again, while we should probably wait on original request instead.

If image is small enough and/or internet connection is fast enough, we end up with loading image 3-4 times.

We use okhttp, do you think this is something that can be fixed by switching glide to use okhttp, or it is issue in glide?

**Comment by danysantiago** [comment]
@vovkab I ran into the same situation and made a post in the GGroup about it: https://groups.google.com/forum/#!topic/glidelibrary/9JPybe6S5UQ

I ended up switching Glide's http stack to Volley. Now when I make multiple requests with the same URL and with SOURCE as disk strategy, Volley only downloads one image and the other requests receive a 304 which then pick the image from cache. I believe OkHttp would work same. Hope this helps!

**Comment by sjudd** [comment]
Thanks for chiming in @danysantiago. 

@vovkab I don't think the requests are being cancelled. If you start multiple requests to Glide in a row with different sizes and/or options, each will make its own http request until the cached source data is present locally. As long as the requests aren't being cancelled (and again I don't think they are), this is WAI (for now).

My thought is that avoiding duplicate network requests is largely a goal of a networking library. I'd like to keep Glide focused on image loading. We've tried to make it easy to integrate any arbitrary networking library, so if you run into a case like this you have options. 

That said, I certainly wouldn't mind improving our default, particularly if others are interested in contributing.

**Comment by vovkab** [comment]
@danysantiago, @sjudd thank you for help

I tried okhttp, but it still trying to load multiple images, looks like it is something else.

I think there is something wrong with this:
```java
        Request previous = target.getRequest();

        if (previous != null) {
            previous.clear();
            requestTracker.removeRequest(previous);
            previous.recycle();
        }
```

As I can see glide is canceling requests, and not http library, so if I set image multiple times in the same view, it will load it from beginning every time on bad connection.

Here is valid use case: if I have a list of images and internet connection is not great, scrolling up and down will reset image loading and start loading again and again.
Maybe there is a way to reconnect to http client requests or do not cancel them, let them finish?

**Comment by TWiStErRob** [comment]
You can't let them queue and finish, you just said they have a bad connection. Imagine someone flings a 100 long list where each images takes more than a second to load. The last one at the end of the list would "never" load.

**Comment by sjudd** [comment]
If you're scrolling back and forth we definitely cancel requests as views go off screen. If you have a limited amount of bandwidth you almost always want to focus on what the user is currently looking at over things they were looking at. There's definitely a worst case scenario where the user just scrolls back and forth continuously and nothing ever loads, but that's not particularly common.

All that said, we don't typically disconnect connections once we've started reading (see [OkHttpStreamFetcher][1] and [HttpUrlFetcher][2], so I'm not sure that this is the cause either. 

[1]: https://github.com/bumptech/glide/blob/master/integration/okhttp/src/main/java/com/bumptech/glide/integration/okhttp/OkHttpStreamFetcher.java
[2]: https://github.com/bumptech/glide/blob/master/library/src/main/java/com/bumptech/glide/load/data/HttpUrlFetcher.java

**Comment by vovkab** [comment]
@TWiStErRob, @sjudd I see your point, yes, list maybe not the best example. 

Here is another one - profile photo in conversation list, it can be there like 3-4 times on screen and then it can get refreshed, which will lead in at least 4 broken attempts of the same image.

I played a little bit with the code and looks like you do guys cancel valid requests, what I mean is "previous" and "new requests" have exactly the same model(url) and exactly the same view.
Maybe we can put a check to see if this view is already processing this model then skip?

**Comment by sjudd** [comment]
Ah yes, so if you're calling notifyDataSetChanged() repeatedly, in a short period of time, requests may be cancelled and restarted repeatedly. 

We probably could help with that by trying to implement equality of some sort on Requests, but it wouldn't be straightforward to do. Even if we did so, we'd still be relying on RecyclerView and/or ListView to provide the same views for the same positions. I'm actually not sure I can think of a great way to solve this. I'm totally open to suggestions?

**Comment by vovkab** [comment]
I think it is better to make it straightforward:
 - if view is already processing this cache key or request, then just skip. 
No need to worry about recycling and etc. This would be a great addition.

Not important at the moment, but something that can be also improved:
Share request: check if request exists, then join, to cancel, remove view from tracker, if no one else is tracking this request - cancel.

One more improvement - timeout before canceling, something like wait for 2 seconds, in case someone reconnect to this request. But this one can be tricky.

Thank you for helping!

**Comment by Gongcong** [comment]
glide cache bitmap in memory that use the key with url & width & height,so you use same url in different size imageview,glide will get the bitmap from disk cache,not the memory,this is my opinion

**Comment by sjudd** [comment]
I actually think these improvements are implemented. Requests are shared internally in Glide if the cache keys match. They can be shared in a networking library if the url matches. We now avoid cancelling/restarting requests if the same request is made into a view any additional times while the original request is still running: #2194

**Comment by shylendramadda** [comment]
> Setting `.diskCacheStrategy(DiskCacheStrategy.SOURCE)` helped to avoid image reloads.
> 
> But I still see image start loading and get dropped pretty fast and start loading again, sometimes 3-4 tries. Sometimes glide start 3-4 request for the same url at the same time, and then some where half the way drops others, sometimes it even loads 2-3 images for the same url.
> 
> But when it finish loading no more drops, looks like this is where `diskCacheStrategy()` kicks in.
> 
> p.s. Is it ok to keep it here or is it better to create a new ticket?

There is no such `DiskCacheStrategy.SOURCE` available in `glide:4.12.0`

**Comment by TWiStErRob** [comment]
@shylendramadda check https://github.com/bumptech/glide/commit/21cad72fe13832e2a3bfb5f679a7ec87811394fa?branch=21cad72fe13832e2a3bfb5f679a7ec87811394fa&diff=split#diff-641504f4c06ef21b84d8a23b638c76a804e570d50840c19049e1161723c45b55 (Scroll to DiskCacheStrategy.java)

**Comment by shylendramadda** [comment]
> @shylendramadda check [21cad72?branch=21cad72fe13832e2a3bfb5f679a7ec87811394fa&diff=split#diff-641504f4c06ef21b84d8a23b638c76a804e570d50840c19049e1161723c45b55](https://github.com/bumptech/glide/commit/21cad72fe13832e2a3bfb5f679a7ec87811394fa?branch=21cad72fe13832e2a3bfb5f679a7ec87811394fa&diff=split#diff-641504f4c06ef21b84d8a23b638c76a804e570d50840c19049e1161723c45b55) (Scroll to DiskCacheStrategy.java)

I tried with all the things like `DATA`, `RESOURCE` `ALL`, `NONE` but no use still blink is there. Actually, I am getting byte streams from the server then I am converting it to bitmap and loading it in the existing image using Glide.

**Comment by mwshubham** [comment]
@shylendramadda  Use `DiskCacheStrategy.RESOURCE` for decoded image `DATA` for original one.


I have a small question regarding duplicate requests. Can we ignore size as a metric when loading an image i.e. two image requests for two target resolutions should fetch the same resource (either from URL or from cache) when Disk cache strategy is set to `RESOURCE`

**Comment by shylendramadda** [comment]
As I said above I used it already.
